keps/sig-apps/0010-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,157 @@
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node.
+* The new pod may not be allocated due to resources taken by other pods.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades.
+
+
+## Goals
+Enable deployment controller to update the containers inside the pods (such as container image, cmd etc.) without tearing down the pods
+
+## Proposal
+
+### API Change
+Augment deployment controller to have a new update strategy type: InPlaceUpdate.
+
+```
+
+type DeploymentStrategy struct {
+	// Type of deployment. Can be "Recreate" or "RollingUpdate" or "InPlaceUpdate". Default is RollingUpdate.
+	// +optional
+	Type DeploymentStrategyType
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update
+type InPlaceUpdateDeployment struct {
+	// The number of pods to be updated in each batch.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
+	// Absolute number is calculated from percentage by rounding down.
+	BatchSize intstr.IntOrString
+}
+```
+
+User specifies the new update strategy type in the spec and issue the request
+
+### Deployment Controller
+
+#### Update
+Below are two alternative approaches for in-place update in simplified words. 
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+##### Approach 1
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+    
+    // Pause the old replica set. This is required for this scenario: if we reduce oldReplicaSet immediately, there's a chance
```

Comment:
replicaset does not support pause today. we could potentially add a pause functionality to pause the control of the controller. (based on my experience with building quite a few operators, pausing functionality is important for controllers that will be used/controlled by another higher level controller.)

Reply:
sounds like a different proposal?

Reply:
also like I said in a different comment, "pausing" in the context of a deployment does not freeze scaling to avoid fighting with hpa and is only meant to freeze upgrades so "pausing" for a replicaset is a totally new concept.

Reply:
> sounds like a different proposal?

yes. if we want to consider this approach we need to have another proposal as mentioned in appraoch1 vs approach2 section below.

Reply:
> also like I said in a different comment, "pausing" in the context of a deployment does not freeze scaling to avoid fighting with hpa and is only meant to freeze upgrades so "pausing" for a replicaset is a totally new concept.

it is a different concept. one is for pausing the control entirely, the other is for pausing a specific controlling action.